src/theory/quantifiers/expr_miner.h

```diff
@@ -0,0 +1,96 @@
+/*********************                                                        */
+/*! \file expr_miner.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief expr_miner
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef __CVC4__THEORY__QUANTIFIERS__EXPRESSION_MINER_H
+#define __CVC4__THEORY__QUANTIFIERS__EXPRESSION_MINER_H
+
+#include <map>
+#include "expr/expr_manager.h"
+#include "expr/node.h"
+#include "smt/smt_engine.h"
+#include "theory/quantifiers/sygus_sampler.h"
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+/** Expression miner
+ *
+ * This is a virtual base class for modules that "mines" certain information
+ * from (enumerated) expressions. This includes:
+ * - candidate rewrite rules (--sygus-rr-synth)
+ */
+class ExprMiner
+{
+ public:
+  ExprMiner() : d_sampler(nullptr) {}
+  virtual ~ExprMiner() {}
+  /** initialize
+   *
+   *
+   */
+  virtual void initialize(const std::vector<Node>& vars,
```

Comment:
Does every `ExprMiner` need a list of variables? If yes, can the list of variables differ between different `ExprMiner`s?

Reply:
The expression miners for a given expression miner manager all have the same variable list.

Reply:
I'm adding this to the documentation of expr_miner_manager.

Reply:
In that case, we should store the list of variables in the ExprMinerManager and then just a pointer/reference to the list in the ExprMiners to avoid copying the list.

Reply:
They would depend on someone else owning the vector for them, which wouldn't be the end of the world. Dereferencing the pointer to the vector is a trade off to an extent, yes. You avoid copying a vector and you consume less memory at the cost of an occasional memory dereference. For now, it's fine.